sphinx/application.py

```diff
@@ -342,35 +347,37 @@ def _init_enumerable_nodes(self):
 
     def build(self, force_all=False, filenames=None):
         # type: (bool, List[unicode]) -> None
-        try:
-            if force_all:
-                self.builder.compile_all_catalogs()
-                self.builder.build_all()
-            elif filenames:
-                self.builder.compile_specific_catalogs(filenames)
-                self.builder.build_specific(filenames)
-            else:
-                self.builder.compile_update_catalogs()
-                self.builder.build_update()
-
-            status = (self.statuscode == 0 and
-                      'succeeded' or 'finished with problems')
-            if self._warncount:
-                logger.info(bold('build %s, %s warning%s.' %
-                                 (status, self._warncount,
-                                  self._warncount != 1 and 's' or '')))
-            else:
-                logger.info(bold('build %s.' % status))
-        except Exception as err:
-            # delete the saved env to force a fresh build next time
-            envfile = path.join(self.doctreedir, ENV_PICKLE_FILENAME)
-            if path.isfile(envfile):
-                os.unlink(envfile)
-            self.emit('build-finished', err)
-            raise
-        else:
-            self.emit('build-finished', None)
-        self.builder.cleanup()
+        for builder in self.builders.values():
+            self.builder = builder
+            try:
+                if force_all:
+                    builder.compile_all_catalogs()
+                    builder.build_all()
+                elif filenames:
+                    builder.compile_specific_catalogs(filenames)
+                    builder.build_specific(filenames)
+                else:
+                    builder.compile_update_catalogs()
+                    builder.build_update()
+
+                status = (self.statuscode == 0 and
+                          'succeeded' or 'finished with problems')
+                if self._warncount:
+                    logger.info(bold('build %s, %s warning%s.' %
+                                     (status, self._warncount,
+                                      self._warncount != 1 and 's' or '')))
+                else:
+                    logger.info(bold('build %s.' % status))
+            except Exception as err:
+                # delete the saved env to force a fresh build next time
+                envfile = path.join(self.doctreedir, ENV_PICKLE_FILENAME)
+                if path.isfile(envfile):
+                    os.unlink(envfile)
+                self.emit('build-finished', err)
+                raise
+            builder.cleanup()
+
+        self.emit('build-finished', None)
```

Comment:
Note that I changed the order of cleanup and build-finished here. Not sure how likely this is to cause state-consistency issues in any extension. I did a spot-check on the HTML builder's cleanup function and it didn't look like trouble. 

I think semantically this event (or perhaps a new ``builder-finished`` event, probably paired with a ``builder-started`` on line 352) would ideally be emitted when each builder completes, but I avoided doing so given the same issue about every downstream subscriber needing an update to handle the additional arg.